### ISSUE TITLE
Missing support for SEM models with single manifest (non-indicator) variables.

### ISSUE THREAD

**Issue by K4tana** [sfcheung/semptools#97]
Hi, I was hoping to use your package for my SEM model which includes single manifest variables in my regressions that are not indicators. I didn't find a way to integrate them into your framework. Did I miss something or is this just not possible yet? If it isn't, this would make the package very powerful. If it is, sorry in advance!

**Comment by sfcheung** [comment]
Thanks a lot for your suggestion. I added preliminary support for exogenous variables to `set_sem_layout()` in version 0.2.9.5, which can be installed by

```r
remotes::install_github("sfcheung/semptools")
```

Just include the observed exogenous variables as one-indicator factors in `indicator_order` and `indicator_factor`. They will be recognized as observed exogenous variables and can be positioned by `factor_layout` as if they are factors. Because they have no indicators, in `factor_point_to`, just use `NA`. This is an example:

```r
library(lavaan)
#> This is lavaan 0.6-11
#> lavaan is FREE software! Please report any bugs.
library(semPlot)
library(semptools)
mod <-
  'f1 =~ x01 + x02 + x03
   f3 =~ x08 + x09 + x10
   f4 =~ x11 + x12 + x13 + x14
   f3 ~ f1 + x04
   f4 ~ f3 + x05'
fit_sem <- lavaan::sem(mod, sem_example)
p <- semPaths(fit_sem, whatLabels="est",
        sizeMan = 5,
        nCharNodes = 0, nCharEdges = 0,
        edge.width = 0.8, node.width = 0.7,
        edge.label.cex = 0.6,
        style = "ram", 
        mar = c(10,10,10,10),
        DoNotPlot = TRUE)
indicator_order  <- c("x04", "x05", "x01", "x02", "x03",
                      "x11", "x12", "x13", "x14", "x08", "x09", "x10")
indicator_factor <- c("x04", "x05", "f1",  "f1",  "f1",
                      "f4",  "f4",  "f4",  "f4",  "f3",  "f3",  "f3")
factor_layout <- matrix(c( "f1",  "f3", "f4",
                          "x04", "x05",  NA), byrow = TRUE, 2, 3)
factor_point_to <- matrix(c("up", "up", "up",
                             NA,   NA,      NA), byrow = TRUE, 2, 3)
indicator_push <- c(f3 = 2, f4 = 1.5)
indicator_spread <- c(f1 = 2, f4 = 1.5)
loading_position <- c(f3 = .8)
p2 <- set_sem_layout(p,
                    indicator_order = indicator_order,
                    indicator_factor = indicator_factor,
                    factor_layout = factor_layout,
                    factor_point_to = factor_point_to,
                    indicator_push = indicator_push,
                    indicator_spread = indicator_spread,
                    loading_position = loading_position)
plot(p2)
```

This is the plot:

![image](https://user-images.githubusercontent.com/7019263/162615288-ceaa5c1c-8597-460f-9779-88b95e7c819e.png)

There are still issues that I do not have solutions yet. For example, as shown in the figure above, the function still assumes that the observed exogenous variables have indicators and so reserves some space for them in the plot. I will think about a better solution to handle observed exogenous variables. I decided to push this change now because this version may already be good enough for some cases. (E.g., the empty space can be used to add information such as model fit measures).

If you do not mind doing things not officially supported, for an example like the one above, we can also manually change the margin to remove the empty area by setting `$plotOptions$mar` using a vector of four numbers (bottom, left, top, and right):

```r
p2_fix <- p2
p2_fix$plotOptions$mar <- c(0, 1, 1, 1)
plot(p2_fix)
```

This is the result:

![image](https://user-images.githubusercontent.com/7019263/162616281-bf61d5b2-d339-402e-8ff2-bf1b473f959a.png)

I have not tested models with observed endogenous variables affected by latent factors because it may not be easy to differentiate them from indicators. A tentative solution is to include them as one-indicator factors and fix the loadings to one and error variances to zero.

I will keep this issue open for now until I have a plan on how to provide better support for observed exogenous variables.

**Comment by K4tana** [comment]
Thanks for the quick reply!

I found a workaround by myself actually, it can be done when one predefines the matrix via semPaths(). I found out that the layout parameter accepts any variable as input as long as it is included in the model. It might be practical to instruct users to define their layout matrix before plotting with semPaths() and then use the semptools functions to modify the plot. I would propose a smoothing function which optimizes the plot. I discovered that the matrix input for the layout parameter in semPaths produces ugly  results here and there, as sometimes indicators are spaced too close to the latent variable and therefore loadings aren’t displayed well or very crooked and arrows aren’t clearly visible. A function that could ensure a minimum distance would be of great help for users. I could imagine a trial and error approach that edits the layout  matrices so that endogenous and exogenous nodes have a specified (maybe modifiable) minimum distance from another (could be Euclidean or something like that).

 Regards,
Oliver D. Reithmaier, B.Sc.

Psychologist @ LMU Munich
Junior Consultant Information Security @ Corporate Trust GmbH <https://www.corporate-trust.de/>
Connect @ Github <https://github.com/K4tana> LinkedIn <https://www.linkedin.com/in/oreithmaier> XING <https://www.xing.com/profile/OliverD_Reithmaier/cv>
